### PR TITLE
fix(dify-agent): harden E2B binding lifecycle

### LINKED ISSUE #41619: Dify Agent binding creation fails on stale E2B HTTP/2 connections
## Summary

Dify Agent binding creation and cleanup can fail when E2B 2.34.0 reuses a stale native HTTP/2 connection. The backend reports:

```text
502 binding_create_failed
Invalid input ConnectionInputs.RECV_PING in state ConnectionState.CLOSED
```

The console then surfaces a generic `completion_request_error`, so an inline Agent's first Build chat appears to fail even though the problem is in the E2B control-plane transport.

Internal incident context: [Linear DIFY-2966](https://linear.app/dify/issue/DIFY-2966/inline-agent-%E5%AF%B9%E8%AF%9D%E5%A4%B1%E8%B4%A5%E7%9A%84%E9%97%AE%E9%A2%98).

## Impact

The affected E2B-backed paths are shared by inline Agents, roster Agents, Workflow Agent V2, Binding files, Home Snapshots, and Binding cleanup. A release-time `pause()` failure can also overwrite an otherwise successful Agent run.

## Upstream evidence

- [E2B Python SDK 2.38.0](https://github.com/e2b-dev/E2B/releases/tag/%40e2b/python-sdk%402.38.0) moves the REST control plane and envd HTTP traffic from native httpx/httpcore/h2 transports to pyqwest/reqwest/hyper.
- [E2B issue #1418](https://github.com/e2b-dev/E2B/issues/1418#issuecomment-5244076939) documents the same class of HTTP/2 closed-state errors; the maintainer recommends upgrading to 2.38.0.

## Expected behavior

- Binding lifecycle operations do not fail with the old `ConnectionState.CLOSED` signature.
- Safe, idempotent E2B control-plane operations receive one bounded retry for transient transport failures.
- Cleanup failures are observable but do not replace a successful Agent result.
- Binding create remains one-shot: no blind retry or automatic adoption after an ambiguous response.

## Proposed change

- Raise the Dify Agent server minimum to E2B 2.38.0 and update locks.
- Harden only the E2B backend's idempotent lifecycle operations and release cleanup.
- Add focused unit coverage and an opt-in staging lifecycle soak test with strict ownership and cleanup gates.

### PR DESCRIPTION
**Pull request by zhaohao1004**
## Summary

- upgrade the Dify Agent server runtime from E2B 2.34.0 to 2.38.0, removing the affected native `httpcore/h2` control-plane path
- retry `connect`, `pause`, `kill`, and `delete_snapshot` once after 250 ms for bounded transport failures, including `ReadTimeout`
- keep Binding create, Home Snapshot create, and file operations one-shot; ambiguous create results are not automatically adopted
- keep ordinary E2B release cleanup failures observable without replacing a successful Agent result or swallowing cancellation
- return provider capacity exhaustion as HTTP 429 `binding_capacity_exhausted` without replaying create
- add focused regression tests and an opt-in staging soak with exact ownership, paused-state checks, redacted JSONL evidence, and zero-residue cleanup

## Root cause and scope

Linear DIFY-2966 was triggered when E2B 2.34.0 reused a stale native HTTP/2 connection and received `ConnectionInputs.RECV_PING` after the h2 state was already `CLOSED`. Dify wrapped the resulting protocol error as `502 binding_create_failed`, which surfaced as an inline Agent chat failure.

This is an E2B backend lifecycle issue, not a frontend or inline-Agent-specific bug. The affected backend is shared by inline Agents, roster Agents, Workflow Agent V2, Binding files, Home Snapshots, and Binding cleanup.

[E2B Python SDK 2.38.0](https://github.com/e2b-dev/E2B/releases/tag/%40e2b/python-sdk%402.38.0) migrates the relevant transport stack. [E2B #1418](https://github.com/e2b-dev/E2B/issues/1418#issuecomment-5244076939) documents the same closed-state failure class.

## Behavior contract

Retried once:

- `connect`
- `pause`
- `kill`
- `delete_snapshot`
- only `ConnectError`, `ReadTimeout`, `ReadError`, `WriteError`, and `RemoteProtocolError`

Not retried:

- Binding create and Home Snapshot create
- whole file operations or file writes
- not-found, HTTP status, validation, and cancellation failures

Release always closes the shellctl data plane before pausing. Ordinary close/pause failures produce structured warnings and do not mask a successful result; `BaseException` cancellation semantics remain unchanged.

Fixes #41619

## Validation

Local gates:

- `cd dify-agent && uv lock --check`
- `cd dify-agent && make check`
- `cd dify-agent && make typecheck`
- `cd dify-agent && env -u ALL_PROXY -u all_proxy -u HTTP_PROXY -u http_proxy -u HTTPS_PROXY -u https_proxy make test` — 848 passed, 26 skipped
- `cd api && uv lock --check`
- independent behavior and exception-boundary review passed

Final candidate provenance:

- source: `09b82cf456536690b760ab177ecdcdabfa37c532`
- build: [Build and Push #33606719717](https://github.com/langgenius/dify/actions/runs/33606719717)
- OCI index: `sha256:b9271bcefd4561d9da23e479dce1de4f746527b52ef0d365daf9a50246fdee1a`
- amd64: `sha256:b192a51c8476af25ce078b2c0e5df70a355809dcb8fe56b1ffbe809042c7f23c`
- arm64: `sha256:62b999f7b6d4f7852f652e970b1f2f42c07d5a3c3b577079ff41d3411b12c2c1`
- pulled arm64 image verified `COMMIT_SHA=09b82cf456536690b760ab177ecdcdabfa37c532` and `e2b=2.38.0`; ordinary build branch skipped E2B Template sync

Staging evidence against the final candidate:

- immediate pause/resume c20 reproduction: 40/40 create, list, read, and destroy; one real E2B `ReadTimeout` triggered the new bounded retry and then succeeded
- two-Pod ClusterIP/failover smoke: 60/60 create, list, read, and destroy; 20/20 paused-state identity checks before and after the idle probe
- deleting one Ready Pod left only the survivor endpoint for the next probe; the survivor completed all 20 list/read operations before the replacement served traffic
- deterministic same-Binding handoff passed in both directions: Pod A create → Pod B list/read/destroy, and Pod B create → Pod A list/read/destroy
- retry-exhausted, release cleanup warning, `RECV_PING ... CLOSED`, `LocalProtocolError`, and `RemoteProtocolError` counts were zero
- controlled Pods had zero restarts; final `dify-2966-*` E2B residue was zero; JSONL ledgers contained no credentials

The final image is running in staging through an explicitly authorized temporary live Deployment override while Argo auto-sync is disabled. Argo is expectedly `OutOfSync/Healthy`; the saas-deploy refresh PR remains draft and unmerged.

The staging key still reports an effective limit of 20. A separate shared-environment attempt demonstrated that one organic running Sandbox can produce provider capacity exhaustion at c20; the final code now returns this as HTTP 429 without retrying create. c20 therefore remains interim correctness/stability evidence, and the original c32 capacity gate has not passed.

## Screenshots

Not applicable; backend/runtime-only change.

## Checklist

- [x] This change does not require a user-facing documentation update.
- [x] The linked issue documents the failure and expected behavior.
- [x] Regression coverage preserves retry, one-shot create, cleanup, cancellation, and capacity boundaries.
- [x] Relevant lint, typecheck, lock, unit-test, image-build, and staging checks were run.

From Codex

### CODE REVIEW

File: dify-agent/src/dify_agent/runtime_backend/e2b.py [hunk 1]
```diff
@@ -9,16 +9,20 @@
 from __future__ import annotations
 
 import asyncio
+from collections.abc import Awaitable, Callable
 from dataclasses import dataclass, field
-from typing import TYPE_CHECKING, Literal, Protocol, cast
+import logging
+from typing import TYPE_CHECKING, Literal, Protocol, TypeVar, cast
 
+import httpx as e2b_httpx
```
Comment (BeautyyuYanli) [inline dify-agent/src/dify_agent/runtime_backend/e2b.py:17]: use httpx2 with a new instance to set args, instead of introducing httpx and setting its module level args

**Review by BeautyyuYanli — CHANGES_REQUESTED** [review]

File: dify-agent/src/dify_agent/runtime_backend/e2b.py (hunk 1 above)
Comment (zhaohao1004) [reply dify-agent/src/dify_agent/runtime_backend/e2b.py:17]: Thanks for the review. One constraint here is that E2B Python SDK 2.38 internally uses native `httpx` through its pyqwest HTTPX transport, and the timeout observed in staging is a native `httpx.ReadTimeout`. `httpx2.ReadTimeout` belongs to a separate exception hierarchy, and E2B’s public `ApiParams` does not expose an `httpx2.AsyncClient`/transport injection point. Replacing these exception types with `httpx2` would therefore stop the bounded retry from catching the actual E2B failure.

This code is not changing native `httpx` module-level settings; the import is only used to classify transport exceptions emitted by the E2B SDK.

Would isolating the native `httpx` catches inside `E2BSDKControlPlane` and translating them into an internal typed control-plane error address your concern? That would keep the rest of the runtime backend on `httpx2` and prevent the E2B transport implementation detail from leaking into the lifecycle adapter. If you had a different E2B client-injection hook in mind, could you point me to it?

**Review by BeautyyuYanli — APPROVED** [review]

**Review by wylswz — APPROVED** [review]
